Examples validation

danicc097 opened the pull request:
First draft to fix #588. Let me know what you think @fenollp before digging more into it and having proper tests.

Should this be non-configurable? Based on the spec: 

> Tooling implementations MAY choose to validate compatibility automatically, and reject the example value(s) if incompatible.`

fenollp commented on openapi3/example.go:
```diff
@@ -55,5 +55,9 @@ func (example *Example) UnmarshalJSON(data []byte) error {
+	if v := example.Value; v == nil && example.ExternalValue == "" {
+		return fmt.Errorf("example has no value field")
+	}
+
```
I'd also add some code that ensures
>  The value field and externalValue field are mutually exclusive.

fenollp commented on openapi3/example_validation_test.go:
```diff
@@ -0,0 +1,268 @@
+	require.Error(t, err)
+}
+
+// XXX should this be a validation error or skipped?
```
This should be an error complaining that either `value` or `externalValue` should be set.

fenollp commented on openapi3/example_validation_test.go:
```diff
@@ -0,0 +1,268 @@
+      type: object
+info:
+  title: An API
+  version: 1.2.3.4
```
I see these docs are mostly the same and are quite large in that it's not easy to notice their differences. Maybe use string concatenation to show the reusable parts and the differing parts.

danicc097 replied:
It certainly was a bad idea to choose YAML for this. Let me know if now it's decent enough or should refactor to JSON/build it from structs

fenollp commented on openapi3/example_validation.go:
```diff
@@ -0,0 +1,18 @@
+)
+
+func ValidateExampleValue(ctx context.Context, input interface{}, schema *Schema) error {
+	opts := make([]SchemaValidationOption, 0, 3) // 3 potential opts here
```
I only see two...

fenollp commented on openapi3/example_validation.go:
```diff
@@ -0,0 +1,18 @@
+
+func ValidateExampleValue(ctx context.Context, input interface{}, schema *Schema) error {
+	opts := make([]SchemaValidationOption, 0, 3) // 3 potential opts here
+	opts = append(opts, VisitAsRequest())
```
Then what about examples in responses?

danicc097 replied:
Very true + added tests for this

fenollp commented on openapi3/media_type.go:
```diff
@@ -80,7 +81,22 @@ func (mediaType *MediaType) Validate(ctx context.Context) error {
+			if err := ValidateExampleValue(ctx, example, schema.Value); err != nil {
+				return err
+			}
+		} else if examples := mediaType.Examples; examples != nil {
```
This needs another check that ensures
>  The example object is mutually exclusive of the examples object.

fenollp commented on openapi3/refs.go:
```diff
@@ -290,6 +291,12 @@ func (value *SchemaRef) UnmarshalJSON(data []byte) error {
 // Validate returns an error if SchemaRef does not comply with the OpenAPI spec.
 func (value *SchemaRef) Validate(ctx context.Context) error {
 	if v := value.Value; v != nil {
+		if x := v.Example; x != nil {
+			if err := ValidateExampleValue(ctx, x, v); err != nil {
+				return fmt.Errorf("invalid schema example: %s", err)
+			}
+		}
+
```
Actually this must go within `func (schema *Schema) Validate(ctx context.Context) error {...}` otherwise validating schemas without dereferencing a `$ref` won't go through this code.

fenollp reviewed (changes requested):
Thanks for taking the time to do this :)

---

> Should this be non-configurable? Based on the spec:
>>    Tooling implementations MAY choose to validate compatibility automatically, and reject the example value(s) if incompatible.`

We recently merged #590 which added https://github.com/getkin/kin-openapi/blob/46603c3b350c28a74763bda34fd4fe2142a76145/openapi3/validation_options.go#L9-L12
so I'd suggest to add an option there with strict defaults.

danicc097 commented:
I think that should cover just about everything (valid request, response and schema examples or disabled validation flag raise no error at all, else it errors out). Thanks for the inputs and if you have any further comments PLMK

fenollp commented on openapi3/media_type.go:
```diff
@@ -73,14 +74,37 @@ func (mediaType *MediaType) UnmarshalJSON(data []byte) error {
 
 // Validate returns an error if MediaType does not comply with the OpenAPI spec.
 func (mediaType *MediaType) Validate(ctx context.Context) error {
+	validationOpts := getValidationOptions(ctx)
```
Think of the hot path. Move this where it's used `if validationOpts := getValidationOptions(ctx); validationOpts.ExamplesValidationDisabled { ...`

fenollp commented on openapi3/example_validation.go:
```diff
@@ -0,0 +1,19 @@
+	"context"
+)
+
+func ValidateExampleValue(ctx context.Context, input interface{}, schema *Schema) error {
```
Why is this public?

fenollp commented on openapi3/example.go:
```diff
@@ -55,5 +56,17 @@ func (example *Example) UnmarshalJSON(data []byte) error {
-	return nil // TODO
+	validationOpts := getValidationOptions(ctx)
+
+	if validationOpts.ExamplesValidationDisabled {
```
Not that kind of validation. The validation of the example against the schema should be opt-out, the validation of the example object should not.

fenollp commented on openapi3/example_validation.go:
```diff
@@ -0,0 +1,19 @@
+
+func ValidateExampleValue(ctx context.Context, input interface{}, schema *Schema) error {
+	opts := make([]SchemaValidationOption, 0, 3) // 3 potential opts here
+	opts = append(opts, VisitAsResponse())
+	opts = append(opts, VisitAsRequest())
```
What are you trying to do with these options here and how do you think this gets tested?

danicc097 commented:
I added parameters validation which I completely forgot and implemented all changes requested. Do you prefer the `mustErr` field  trick for example object validation or duplicate to its own `TestExamplesValidation` and rename the rest to `TestExamplesSchemaValidation` or something like that?

fenollp commented on openapi3/example_validation_test.go:
```diff
@@ -8,22 +8,51 @@ import (
+		parametersExample       string
+		componentExamples       string
+		errContains             string
+		mustErr                 bool
```
To me `mustErr` is redundant with `errContains`: `mustErr = errContains != ""`.

fenollp commented on openapi3/parameter.go:
```diff
@@ -307,6 +307,26 @@ func (parameter *Parameter) Validate(ctx context.Context) error {
 			return fmt.Errorf("parameter %q schema is invalid: %v", parameter.Name, err)
 		}
+		if parameter.Example != nil && parameter.Examples != nil {
+			return fmt.Errorf("%s: example and examples are mutually exclusive", schema.Ref)
```
```suggestion
			return fmt.Errorf("parameter %q example and examples are mutually exclusive", parameter.Name)
```

fenollp commented on openapi3/media_type.go:
```diff
@@ -80,7 +81,28 @@ func (mediaType *MediaType) Validate(ctx context.Context) error {
 			return err
 		}
+		if mediaType.Example != nil && mediaType.Examples != nil {
+			return fmt.Errorf("%s: example and examples are mutually exclusive", schema.Ref)
```
`schema.Ref` can be empty. You'll have to use something else :)

fenollp commented on openapi3/parameter.go:
```diff
@@ -307,6 +307,26 @@ func (parameter *Parameter) Validate(ctx context.Context) error {
+			return fmt.Errorf("%s: example and examples are mutually exclusive", schema.Ref)
+		}
+		if validationOpts := getValidationOptions(ctx); validationOpts.ExamplesValidationDisabled {
+			return nil
```
Can't return here. There's non-Example code below this block...

danicc097 commented on openapi3/example_validation_test.go:
```diff
@@ -0,0 +1,426 @@
+	}
+}
+
+func TestExampleObjectValidation(t *testing.T) {
```
to get rid of `mustErr` and make intent clearer

fenollp commented on openapi3/example_validation.go:
```diff
@@ -0,0 +1,9 @@
+package openapi3
+
+import (
+	"context"
+)
+
+func validateExampleValue(ctx context.Context, input interface{}, schema *Schema) error {
```
```suggestion
func validateExampleValue(input interface{}, schema *Schema) error {
```

fenollp commented on openapi3/media_type.go:
```diff
@@ -80,7 +82,28 @@ func (mediaType *MediaType) Validate(ctx context.Context) error {
+			return nil
+		}
+		if example := mediaType.Example; example != nil {
+			if err := validateExampleValue(ctx, example, schema.Value); err != nil {
```
```suggestion
			if err := validateExampleValue(example, schema.Value); err != nil {
```

fenollp commented on openapi3/media_type.go:
```diff
@@ -80,7 +82,28 @@ func (mediaType *MediaType) Validate(ctx context.Context) error {
+				if err := v.Validate(ctx); err != nil {
+					return fmt.Errorf("%s: %s", k, err)
+				}
+				if err := validateExampleValue(ctx, v.Value.Value, schema.Value); err != nil {
```
```suggestion
				if err := validateExampleValue(v.Value.Value, schema.Value); err != nil {
```

fenollp commented on openapi3/parameter.go:
```diff
@@ -303,16 +303,38 @@ func (parameter *Parameter) Validate(ctx context.Context) error {
+			return nil
+		}
+		if example := parameter.Example; example != nil {
+			if err := validateExampleValue(ctx, example, schema.Value); err != nil {
```
```suggestion
			if err := validateExampleValue(example, schema.Value); err != nil {
```

fenollp commented on openapi3/parameter.go:
```diff
@@ -303,16 +303,38 @@ func (parameter *Parameter) Validate(ctx context.Context) error {
+				if err := v.Validate(ctx); err != nil {
+					return fmt.Errorf("%s: %s", k, err)
+				}
+				if err := validateExampleValue(ctx, v.Value.Value, schema.Value); err != nil {
```
```suggestion
				if err := validateExampleValue(v.Value.Value, schema.Value); err != nil {
```

fenollp commented on openapi3/schema.go:
```diff
@@ -753,6 +753,12 @@ func (schema *Schema) validate(ctx context.Context, stack []*Schema) (err error)
 	}
 
+	if x := schema.Example; x != nil && !validationOpts.ExamplesValidationDisabled {
+		if err := validateExampleValue(ctx, x, schema); err != nil {
```
```suggestion
		if err := validateExampleValue(x, schema); err != nil {
```

fenollp commented:
Thanks a lot for this work! 

I note support for `readOnly` and `writeOnly` schema fields is missing (see [spec the here](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.0.md#fixed-fields-20)). This is where `VisitAsRequest()` should be employed when validating examples that go in requests and respectively `VisitAsResponse()` responses. Let's do this as another PR!
